server/src/main/java/org/opensearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -261,11 +261,24 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
                 // advances and not when creating a new safe commit. In any case this is a best-effort thing since future recoveries can
                 // always fall back to file-based ones, and only really presents a problem if this primary fails before things have settled
                 // down.
-                startingSeqNo = softDeletesEnabled
-                    ? Long.parseLong(safeCommitRef.getIndexCommit().getUserData().get(SequenceNumbers.LOCAL_CHECKPOINT_KEY)) + 1L
-                    : 0;
+                if(softDeletesEnabled) {
+                    final long minimumRetainingSequenceNumber = shard.getRetentionLeases()
+                        .leases()
+                        .stream()
+                        .mapToLong(RetentionLease::retainingSequenceNumber)
+                        .min()
+                        .orElse(Long.MAX_VALUE);
+                    final long safeCommitSeqNo = Long.parseLong(safeCommitRef.getIndexCommit()
```

Comment:
If I understand this right, tlog recovery is a mechanism to optimize serialization. Why would start sequence number change?

Reply:
We want to have the same set of tlog ops from the primary shard onto the replica shard during the peer recovery. This ensures the ops are available in tlog as well in newly recovered shard